Load CSV files fails on date and numeric types

### Describe the bug
When you try to seed csv into (hive) it fails if all the types are string.

### Steps To Reproduce
1. Install dbt 0.18.1 and dbt-spark
2. create a simple csv file as below:
```csv
id,name,first_payment_date,maturity_date,loan_amount,installment_amount
1,LAI-1,2020-01-01,2020-10-01,100000.00,1000.00
```
3. Define some column types for your csv
```yml
seeds:
  my_new_project:
    loan_account:
      +column_types:
        id: string
        name: string
        first_payment_date: timestamp
        maturity_date: timestamp
        loan_amount: decimal(20,2)
        installment_amount: decimal(20,2)
```


### Expected behavior
Should be able to insert csv data easily

If you specify a column_type as double / decimal in column_types, the csv load treats them as string
https://github.com/fishtown-analytics/dbt/blob/v0.18.1/core/dbt/context/providers.py#L732

I want loan_amount to be treated as a decimal. If I omit it from the `+column_types` then it gets treated as double.
There doesn't seem to be any easy way to override the csv loading behaviour.


### Screenshots and log output
```
Completed with 1 error and 0 warnings:

Runtime Error in seed loan_account (data/loan_account.csv)
  Database Error
    Error running query: org.apache.spark.sql.AnalysisException: Cannot write incompatible data to table '`default`.`loan_account`':
    - Cannot safely cast 'loan_amount': string to decimal(20,2)
    - Cannot safely cast 'installment_amount': string to decimal(20,2);
```

The SQL that is generated is 
```sql
insert into default.loan_account values('1','LAI-1','2020-01-01', '2020-01-10','100', '100')
```

The desired output should be something like this
```sql
insert into default.loan_account values('1','LAI-1',cast('2020-01-01' as Date), cast('2020-01-10' as Date),100, 100)
```

1. Wrap the timestamp / dates in CAST
2. and don't treat numeric values as strings



## Additional info

I've customised a the macro a little to support cast as date, but its messy.
```sql
{% macro spark__load_csv_rows(model, agate_table) %}
    {% set batch_size = 1000 %}

    {% set statements = [] %}

    {% for chunk in agate_table.rows | batch(batch_size) %}
        {% set bindings = [] %}

        {% for row in chunk %}
          {% do bindings.extend(row) %}
        {% endfor %}

        {% set sql %}
            insert into {{ this.render() }} values
            {% for row in chunk -%}
                ({%- for column in agate_table.columns -%}
                    {%- if 'ISODate' in (column.data_type | string) -%}
                      cast(%s as date)
                    {%- elif 'date' in (column.name) -%}
                      cast(%s as date)
                    {%- else -%}
                    %s
                    {%- endif -%}
                    {%- if not loop.last%},{%- endif %}
                {%- endfor -%})
                {%- if not loop.last%},{%- endif %}
            {%- endfor %}
        {% endset %}

        {% for column in agate_table.columns  -%}
          {% do log(column.name ~ (column.data_type | string), True) %}
        {%- endfor %}

        {% do log(sql, True) %}
        {% do adapter.add_query(sql, bindings=bindings, abridge_sql_log=True) %}

        {% if loop.index0 == 0 %}
            {% do statements.append(sql) %}
        {% endif %}
    {% endfor %}

    {# Return SQL so we can render it out into the compiled files #}
    {{ return(statements[0]) }}
{% endmacro %}
```

@jtcohen6 I'm trying to contribute this fix back into the repository. 

I'm starting by creating a unit test the seed macros:
- spark__load_csv_rows
- spark__reset_csv_table
- spark__create_csv_table

I've created this test class:

```python
# test_seed_macros

import unittest
from unittest import mock
import os
import re
from jinja2 import Environment, FileSystemLoader
from dbt.clients import agate_helper
from dbt.clients.jinja import MaterializationExtension
from tempfile import mkdtemp

SAMPLE_CSV_DATA = """a,b,c,d,e,f,g
1,n,test,3.2,20180806T11:33:29.320Z,True,NULL
2,y,asdf,900,20180806T11:35:29.320Z,False,a string"""

class TestSparkSeedMacros(unittest.TestCase):

    def setUp(self):
        self.jinja_env = Environment(loader=FileSystemLoader(
                                        ['dbt/include/spark/macros',
                                         'dbt/include/spark/macros/materializations']),
                                     extensions=['jinja2.ext.do', MaterializationExtension])

        self.config = {}
        self.default_context = {
            'validation': mock.Mock(),
            'model': mock.Mock(),
            'exceptions': mock.Mock(),
            'config': mock.Mock()
        }
        self.default_context['config'].get = lambda key, default=None, **kwargs: self.config.get(key, default)
        self.tempdir = mkdtemp()

    def __get_template(self, template_filename):
        return self.jinja_env.get_template(template_filename, globals=self.default_context)

    def __run_macro(self, template, name, agate_table):
        self.default_context['model'].alias = 'test'
        value = getattr(template.module, name)(self.default_context['model'], agate_table)
        return re.sub(r'\s\s+', ' ', value)

    def test_macros_load(self):
        self.jinja_env.get_template('seed.sql')

    def test_macros_create_csv_table(self):
        template = self.__get_template('seed.sql')

        path = os.path.join(self.tempdir, 'input.csv')
        with open(path, 'wb') as fp:
            fp.write(SAMPLE_CSV_DATA.encode('utf-8'))

        agate_table = agate_helper.from_csv(path, [])
        sql = self.__run_macro(template, 'spark__create_csv_table', agate_table).strip()

        self.assertEqual(sql, "create table my_table as select 1")
```

When I run this code I get the following error:

```
self = Undefined, args = (), kwargs = {}

    @internalcode
    def _fail_with_undefined_error(self, *args, **kwargs):
        """Raise an :exc:`UndefinedError` when operations are performed
        on the undefined value.
        """
>       raise self._undefined_exception(self._undefined_message)
E       jinja2.exceptions.UndefinedError: 'this' is undefined

.tox/unit/lib/python3.8/site-packages/jinja2/runtime.py:747: UndefinedError
```

I assume the following snippet in `seed.sql` which is missing.
```
create table {{ this.render() }}
```

I think ideally I would just want integration tests that tests the seed on spark 2.4.x and spark 3.0.0 but I'm not familiar with circle CI and it seems tricky to setup locally.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,11 +4,10 @@
 - Allow setting table `OPTIONS` using `config` ([#171](https://github.com/fishtown-analytics/dbt-spark/pull/171))
 
 ### Fixes
-
 - Cast `table_owner` to string to avoid errors generating docs ([#158](https://github.com/fishtown-analytics/dbt-spark/pull/158), [#159](https://github.com/fishtown-analytics/dbt-spark/pull/159))
+- Explicitly cast column types when inserting seeds ([#139](https://github.com/fishtown-analytics/dbt-spark/pull/139), [#166](https://github.com/fishtown-analytics/dbt-spark/pull/166))
 
 ### Under the hood
-
 - Parse information returned by `list_relations_without_caching` macro to speed up catalog generation ([#93](https://github.com/fishtown-analytics/dbt-spark/issues/93), [#160](https://github.com/fishtown-analytics/dbt-spark/pull/160))
 - More flexible host passing, https:// can be omitted ([#153](https://github.com/fishtown-analytics/dbt-spark/issues/153))
 

--- a/dbt/include/spark/macros/materializations/seed.sql
+++ b/dbt/include/spark/macros/materializations/seed.sql
@@ -1,6 +1,7 @@
 {% macro spark__load_csv_rows(model, agate_table) %}
     {% set batch_size = 1000 %}
-
+    {% set column_override = model['config'].get('column_types', {}) %}
+    
     {% set statements = [] %}
 
     {% for chunk in agate_table.rows | batch(batch_size) %}
@@ -13,12 +14,10 @@
         {% set sql %}
             insert into {{ this.render() }} values
             {% for row in chunk -%}
-                ({%- for column in agate_table.columns -%}
-                    {%- if 'ISODate' in (column.data_type | string) -%}
-                      cast(%s as timestamp)
-                    {%- else -%}
-                    %s
-                    {%- endif -%}
+                ({%- for col_name in agate_table.column_names -%}
+                    {%- set inferred_type = adapter.convert_type(agate_table, loop.index0) -%}
+                    {%- set type = column_override.get(col_name, inferred_type) -%}
+                      cast(%s as {{type}})
                     {%- if not loop.last%},{%- endif %}
                 {%- endfor -%})
                 {%- if not loop.last%},{%- endif %}

--- a/test/custom/seed_column_types/data/payments.csv
+++ b/test/custom/seed_column_types/data/payments.csv
@@ -1,0 +1,11 @@
+ID,ORDERID,PAYMENTMETHOD,STATUS,AMOUNT,AMOUNT_USD,CREATED
+1,1,credit_card,success,1000,10.00,2018-01-01
+2,2,credit_card,success,2000,20.00,2018-01-02
+3,3,coupon,success,100,1.00,2018-01-04
+4,4,coupon,success,2500,25.00,2018-01-05
+5,5,bank_transfer,fail,1700,17.00,2018-01-05
+6,5,bank_transfer,success,1700,17.00,2018-01-05
+7,6,credit_card,success,600,6.00,2018-01-07
+8,7,credit_card,success,1600,16.00,2018-01-09
+9,8,credit_card,success,2300,23.00,2018-01-11
+10,9,gift_card,success,2300,23.00,2018-01-12

--- a/test/custom/seed_column_types/test_seed_column_types.py
+++ b/test/custom/seed_column_types/test_seed_column_types.py
@@ -1,0 +1,36 @@
+from cProfile import run
+from test.custom.base import DBTSparkIntegrationTest, use_profile
+import dbt.exceptions
+
+
+class TestSeedColumnTypeCast(DBTSparkIntegrationTest):
+    @property
+    def schema(self):
+        return "seed_column_types"
+        
+    @property
+    def models(self):
+        return "models"
+
+    @property
+    def project_config(self):
+        return {
+            'seeds': {
+                'quote_columns': False,
+            },
+        }
+
+    # runs on Spark v2.0
+    @use_profile("apache_spark")
+    def test_seed_column_types_apache_spark(self):
+        self.run_dbt(["seed"])
+
+    # runs on Spark v3.0
+    @use_profile("databricks_cluster")
+    def test_seed_column_types_databricks_cluster(self):
+        self.run_dbt(["seed"])
+
+    # runs on Spark v3.0
+    @use_profile("databricks_sql_endpoint")
+    def test_seed_column_types_databricks_sql_endpoint(self):
+        self.run_dbt(["seed"])

